scripts/enumerable.rb

```diff
@@ -0,0 +1,238 @@
+# frozen_string_literal: true
+
+module Enumerable
+  def my_each
+    return return_enum unless block_given?
+
+    i = 0
+    while i < size
+      yield(self[i])
+      i += 1
+    end
+  end
+
+  def my_each_with_index
+    return return_enum unless block_given?
+
+    i = 0
+    while i < size
+      yield(self[i], i)
+      i += 1
+    end
+  end
+
+  def my_select
+    return return_enum unless block_given?
+
+    select = []
+    my_each { |item| select << item if yield item }
+    select
+  end
+
+  def my_all?(pattern = nil)
+    all = true
+    # rubocop:disable Style/CaseEquality
+    if block_given?
+      my_each { |item| all = false unless yield item }
+    elsif pattern
+      my_each { |item| all = false unless pattern === item }
+    else
+      my_each { |item| all = false unless item }
+    end
+    # rubocop:enable Style/CaseEquality
+    all
+  end
+
+  def my_any?(pattern = nil)
+    any = false
+    # rubocop:disable Style/CaseEquality
+    if block_given?
+      my_each { |item| any = true if yield item }
+    elsif pattern
+      my_each { |item| any = true if pattern === item }
+    else
+      my_each { |item| any = true if item }
+    end
+    # rubocop:enable Style/CaseEquality
+    any
+  end
+
+  def my_none?(pattern = nil)
+    none = true
+    # rubocop:disable Style/CaseEquality
+    if block_given?
+      my_each { |item| none = false if yield item }
+    elsif pattern
+      my_each { |item| none = false if pattern === item }
+    else
+      my_each { |item| none = false if item }
+    end
+    # rubocop:enable Style/CaseEquality
+    none
+  end
+
+  def my_count(compare = nil)
+    count = 0
+    if block_given?
+      my_each { |item| count += 1 if yield item }
+    elsif !compare.nil?
+      my_each { |item| count += 1 if item == compare }
+    else
+      each { |_item| count += 1 }
+    end
+    count
+  end
+
+  def my_map(my_proc = nil)
+    map = []
+    if my_proc
+      my_proc.call
+    elsif block_given?
+      each { |item| map << yield(item) }
+    else
+      return to_enum
+    end
+    map
+  end
+
+  def my_inject(sym = nil, memo = self[0])
+    raise 'No block given' unless block_given? || !sym.nil?
+    # rubocop:disable Style/CaseEquality
+    if sym
+      my_each do |item|
+        next if item === self[0]
```

Comment:
I don't think this method is efficient for cases where the first two values are identical. See below:

```ruby
irb(main):017:0> [2,2,2,2,4,5,6,7,8,9,10].inject(:+)
=> 57
irb(main):018:0> [2,2,2,2,4,5,6,7,8,9,10].my_inject(:+)
=> 51
irb(main):021:0> [2,2,4,5,6].inject{ |sum, n| sum + n }
=> 19
irb(main):022:0> [2,2,4,5,6].my_inject{ |sum, n| sum + n }
=> 17
```
I would suggest handling cases of how arguments are passed differently.